kvbc/src/categorization/kv_blockchain.cpp

```diff
@@ -429,46 +497,54 @@ void KeyValueBlockchain::deleteLastReachableBlock(BlockId block_id,
                                                   const BlockMerkleOutput& updates_info,
                                                   storage::rocksdb::NativeWriteBatch& batch) {
   merkle_num_of_deleted_keys_.Get().Inc(updates_info.keys.size());
-  std::get<detail::BlockMerkleCategory>(getCategory(category_id))
+  std::get<detail::BlockMerkleCategory>(getCategoryRef(category_id))
       .deleteLastReachableBlock(block_id, updates_info, batch);
 }
 
 // Updates per category
 
-bool KeyValueBlockchain::insertCategoryMapping(const std::string& cat_id,
-                                               const CATEGORY_TYPE type,
-                                               concord::storage::rocksdb::NativeWriteBatch& write_batch) {
+void KeyValueBlockchain::insertCategoryMapping(const std::string& cat_id, const CATEGORY_TYPE type) {
   // check if we know this category type already
-  if (categorires_types_.count(cat_id) == 1) {
-    return false;
+  ConcordAssertEQ(category_types_.count(cat_id), 0);
+
+  // cache the type in memory and store it in DB
+  const auto inserted = category_types_.try_emplace(cat_id, type).second;
+  if (!inserted) {
+    LOG_FATAL(CAT_BLOCK_LOG, "Category [" << cat_id << "] already exists in type map");
+    ConcordAssert(false);
+  }
+  native_client_->put(detail::CAT_ID_TYPE_CF, cat_id, std::string(1, static_cast<char>(type)));
+}
+
+void KeyValueBlockchain::addNewCategory(const std::string& cat_id, CATEGORY_TYPE type) {
+  insertCategoryMapping(cat_id, type);
+  auto inserted = false;
+  switch (type) {
+    case CATEGORY_TYPE::block_merkle:
+      inserted = categories_.try_emplace(cat_id, detail::BlockMerkleCategory{native_client_}).second;
+      break;
+    case CATEGORY_TYPE::immutable:
+      inserted = categories_.try_emplace(cat_id, detail::ImmutableKeyValueCategory{cat_id, native_client_}).second;
+      break;
+    case CATEGORY_TYPE::versioned_kv:
+      inserted = categories_.try_emplace(cat_id, detail::VersionedKeyValueCategory{cat_id, native_client_}).second;
+      break;
+    default:
+      ConcordAssert(false);
   }
-  // new category
-  // cache the type in mem and store it in db.
-  if (const auto [itr, inserted] = categorires_types_.try_emplace(cat_id, type); !inserted) {
-    (void)itr;
-    throw std::runtime_error{"Failed to insert new category"};
+  if (!inserted) {
+    LOG_FATAL(CAT_BLOCK_LOG, "Category [" << cat_id << "] already exists in categories map");
+    ConcordAssert(false);
   }
-  write_batch.put(detail::CAT_ID_TYPE_CF, cat_id, std::string(1, static_cast<char>(type)));
-  return true;
 }
 
 BlockMerkleOutput KeyValueBlockchain::handleCategoryUpdates(BlockId block_id,
```

Comment:
Do we actually use these `handleCategoryUpdates` methods anywhere?

Reply:
Yes, we do - when adding a block. I guess they can be rewritten as a single template method, introducing a type map from input to category.